best-practices/environments.md

```diff
@@ -0,0 +1,20 @@
+# Environments
+
+This document clarifies the terminology we use to talk about deployment & testing environments. 
+
+Note that there are at least two separate needs from non-production environments:
+* testing software: as we are developing software, we need a way to appropriately test and approve code & configuration before we deploy it to production.
+* testing data: we have a variety of situations where bulk loading of data requires that the data itself be vetted before it is loaded or changed in production.
+
+Sometimes these two different needs can share the same environment, but when software changes would disrupt the need for data vetting, this can become problematic.
+
+This document is focused on environments for testing **software**.
+
+<dl>
+  <dt>Production (or "prod") </dt>
+  <dd>The production environment is the one targeted for users and production data. These environments tend to be the most robustly provisioned, and also to hold data that we care about, <em>e.g.</em>, that gets preserved to SDR and backed up regularly.</dd>
+  <dt>Staging (or "stage") </dt>
+  <dd>The staging environment mimics production closely, and serves as a last step in testing a feature or set of features against production-like data after the work has undergone initial (user acceptance/quality assurance) testing and it has been merged.</dd>
+  <dt>User Acceptance Testing ("UAT") & Quality Assurance ("QA")</dt>
+  <dd>UAT and QA environments are used to test a feature or set of features early on, as part of development work. These environments may or may not have production-like data. We tend to use UAT environments for sites needing testing beyond the department, and QA for internal testing.</dd>
```

Comment:
This is a subtle point, but some of the work that happens in these environments (both stage and UAT/QA) is not strictly tied to development work. These environments may be necessary for service team members, data managers, designers, etc. to do their daily work outside of dev cycles. Maintaining these environments (deploying dependency updates, keeping IP address lists and certs up to date, etc.) should be understood as a shared responsibility between devs and Ops that facilitates cross-departmental work.

Reply:
@camillevilla okie dokie. I'll take a hack at that and ping you back.

Reply:
I incorporated some of the text from your comment, and omitted the following:

> Maintaining these environments should be understood as a shared responsibility between devs and Ops that facilitates cross-departmental work.

I did that because I (perhaps wrongly) concluded that this is true for all of the environments, and it made the QA/UAT section read heavier than the other ones. Is that OK or would you like me to add it back?

Reply:
¯\\_(ツ)_/¯ It's fine without. My comment was in response to some historical (pre-weekly update script) tendencies. I understand this document is setting up an intentional practice moving forward.